commit/plugin_next.go

```diff
@@ -0,0 +1,355 @@
+	"github.com/smartcontractkit/chainlink-ccip/pkg/logutil"
+	cciptypes "github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+)
+
+// observationNext contains new Observation logic with breaking changes that should replace the existing
+// Observation in the next release.
+func (p *Plugin) observationNext(
+	ctx context.Context, outCtx ocr3types.OutcomeContext, q types.Query,
+) (types.Observation, error) {
+	// Ensure that sequence number is in the context for consumption by all
+	// downstream processors and the ccip reader.
+	ctx, lggr := logutil.WithOCRInfo(ctx, p.lggr, outCtx.SeqNr, logutil.PhaseObservation)
+
+	var discoveryObs dt.Observation
+	var err error
+
+	if p.discoveryProcessor != nil {
+		tStart := time.Now()
+		discoveryObs, err = p.discoveryProcessor.Observation(ctx, dt.Outcome{}, dt.Query{})
+		lggr.Debugw("commit discovery observation finished",
+			"duration", time.Since(tStart), "err", err)
+		if err != nil {
+			lggr.Errorw("failed to discover contracts", "err", err)
+		}
+	}
+
+	// If the contracts are not initialized then only submit contracts discovery related observation.
+	if !p.contractsInitialized.Load() && p.discoveryProcessor != nil {
+		obs := committypes.Observation{DiscoveryObs: discoveryObs}
+		encoded, err := p.ocrTypeCodec.EncodeObservation(obs)
+		if err != nil {
+			return nil, fmt.Errorf("encode discovery observation: %w, observation: %+v", err, obs)
+		}
+
+		lggr.Infow("contracts not initialized, only making discovery observations", "discoveryObs", discoveryObs)
+		lggr.Infow("commit plugin making observation", "encodedObservation", encoded, "observation", obs)
+
+		return encoded, nil
+	}
+
+	prevOutcome, err := p.ocrTypeCodec.DecodeOutcome(outCtx.PreviousOutcome)
+	if err != nil {
+		return nil, fmt.Errorf("decode previous outcome: %w", err)
+	}
+
+	decodedQ, err := p.ocrTypeCodec.DecodeQuery(q)
+	if err != nil {
+		return nil, fmt.Errorf("decode query: %w", err)
+	}
+
+	merkleRootObs, err := p.merkleRootProcessor.Observation(ctx, prevOutcome.MerkleRootOutcome, decodedQ.MerkleRootQuery)
+	if err != nil {
+		lggr.Errorw("get merkle root processor observation",
+			"err", err,
+			"prevMerkleRootOutcome", prevOutcome.MerkleRootOutcome,
+			"decodedQ", decodedQ.MerkleRootQuery,
+		)
+	}
+
+	obs := committypes.Observation{
+		MerkleRootObs:         merkleRootObs,
+		TokenPriceObs:         tokenprice.Observation{},
+		DiscoveryObs:          discoveryObs,
+		ChainFeeObs:           chainfee.Observation{},
+		FChain:                p.ObserveFChain(lggr),
+		OnChainPriceOcrSeqNum: 0,
+	}
+
+	inflightPricesExist := prevOutcome.MainOutcome.InflightPriceOcrSequenceNumber > 0
+	switch inflightPricesExist {
+	case true:
+		// If we have inflight prices destination chain supporting oracles only observe the onchain price ocr seq num.
+		// We use this observation to check if prices are still inflight within the Outcome.
+		obs.OnChainPriceOcrSeqNum, err = p.observeOnChainPriceOcrSeqNum(ctx)
```

Comment:
Is it safe to do it like this? I mean, it would be safer to not update `obs.OnChainPriceOcrSeqNum` when err is returned. That would prevent us from accidentally updating that field with some rubbish returned together with an error


```go
if priceOcrSeqNum, err := err = p.observeOnChainPriceOcrSeqNum(ctx); if err != nil {
     lggr.Errorw("failed to observe on-chain price seq number", "err", err)
} else {
     obs.OnChainPriceOcrSeqNum = priceOcrSeqNum
}
```

Reply:
tl;dr avoid assigning the result unless it's error-free